common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -418,6 +415,47 @@ static boolean equals(byte[] bytes1, int startPos1, byte[] bytes2, int startPos2
         }
     }
 
+    static int equalsConstantTime(byte[] bytes1, int startPos1, byte[] bytes2, int startPos2, int length) {
+        long result = 0;
+        final long baseOffset1 = BYTE_ARRAY_BASE_OFFSET + startPos1;
+        final long baseOffset2 = BYTE_ARRAY_BASE_OFFSET + startPos2;
+        final int remainingBytes = length & 7;
+        final long end = baseOffset1 + remainingBytes;
+        for (long i = baseOffset1 - 8 + length, j = baseOffset2 - 8 + length; i >= end; i -= 8, j -= 8) {
+            result |= UNSAFE.getLong(bytes1, i) ^ UNSAFE.getLong(bytes2, j);
```

Comment:
I think array contents on hotspot 64-bit intel should by default be 8 byte aligned. Due to compressed oops the array object is 8 byte aligned, followed by a 12 byte object header and the 4 byte array length, followed by array contents 😄 
> Since java has a compacting GC, any pointers to data are an extra level of indirection away right?

afaik the references to an object are updated when moved. so an array variable holds a pointer to the array contents (minus some offset for the object header + size).

That being said, apparently you can get the address of an object/array [somehow]( http://hg.openjdk.java.net/code-tools/jol/file/30c3fbadf049/jol-samples/src/main/java/org/openjdk/jol/samples/JOLSample_17_Allocation.java#l64) (best guess via Unsafe)

However, since java GC can relocate an object between getting the address and doing the comparison, you might never know whether it's the right address at the time of comparison :-).

I think the whole complexity might not be worth it, especially since for headers we are likely not comparing 4K buffers but a few dozen bytes at a time.